### PR TITLE
Add cgroupsv2 for getCGroupLimitedCPUCores

### PR DESCRIPTION
**Pull request by Felixoid**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
As a side quest of #51180, here's the test for getCGroupLimitedCPUCores in old containers

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
This is an automated comment for commit 5fdb19672bc443222560df00e0b76add05a1ad10 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51478/5fdb19672bc443222560df00e0b76add05a1ad10/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Common/getNumberOfPhysicalCPUCores.cpp [hunk 1]
```diff
@@ -31,10 +34,89 @@ int32_t readFrom(const char * filename, int default_value)
...
+                if (cpu_limit_str != "max" && cpu_period != 0)
+                {
+                    float cpu_limit = std::stof(cpu_limit_str);
+                    quota_count = std::min(static_cast<uint32_t>(ceil(cpu_limit / cpu_period)), quota_count);
+                }
+            }
+        }
+        current_cgroup = prefix / cgroup_name;
+        // Looking for cpuset.cpus.effective in directories from the current cgroup to the top level
+        while (current_cgroup != prefix.parent_path())
```
Comment (serxa) [inline src/Common/getNumberOfPhysicalCPUCores.cpp:80]: Do we really need this cycle if we already use `.effective`?

**Review by serxa — COMMENTED** [review]
It would be great to also have a simplest possible test

File: src/Common/getNumberOfPhysicalCPUCores.cpp [hunk 2]
```diff
@@ -31,10 +34,89 @@ int32_t readFrom(const char * filename, int default_value)
...
+    std::ifstream contr_file(prefix / "cgroup.controllers");
+    if (contr_file.is_open())
+    {
+        /// First, we identify the cgroup the process belongs
+        std::ifstream cgroup_name_file("/proc/self/cgroup");
+        if (!cgroup_name_file.is_open())
+            return default_cpu_count;
+
+        // cgroup_name_file always starts with '0::' for v2
+        cgroup_name_file.ignore(3);
```
[the remark this answers is not in the document]
Comment (Felixoid) [reply src/Common/getNumberOfPhysicalCPUCores.cpp:48]: Yes, it's a good catch

File: src/Common/getNumberOfPhysicalCPUCores.cpp (hunk 1 above)
Comment (Felixoid) [reply src/Common/getNumberOfPhysicalCPUCores.cpp:80]: yes, it's necessary

```
start_dir=/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/app.slice/app-org.wezfurlong.wezterm-084ffa47df3f4bb692567dedc8b856e4.scope
while :; do ls ${start_dir}/cpuset.cpus.effective && break || start_dir=$(readlink -f "${start_dir}"/..)
done
ls: cannot access '/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/app.slice/app-org.wezfurlong.wezterm-084ffa47df3f4bb692567dedc8b856e4.scope/cpuset.cpus.effective': No such file or directory
ls: cannot access '/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/app.slice/cpuset.cpus.effective': No such file or directory
ls: cannot access '/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/cpuset.cpus.effective': No such file or directory
ls: cannot access '/sys/fs/cgroup/user.slice/user-1000.slice/cpuset.cpus.effective': No such file or directory
/sys/fs/cgroup/user.slice/cpuset.cpus.effective
```

There are no guarantees that the file exists in the `/sys/fs/cgroup/$(/proc/self/cgroup)` directory

**Comment by Felixoid** [comment]
We have a test https://github.com/ClickHouse/ClickHouse/blob/master/tests/integration/test_cgroup_limit/test.py#L41-L50. But it depends on the integration tests environment, unfortunately

File: src/Common/getNumberOfPhysicalCPUCores.cpp (hunk 1 above)
Comment (serxa) [reply src/Common/getNumberOfPhysicalCPUCores.cpp:80]: Ah, we `break` after the first match. I see...

**Comment by Felixoid** [comment]
The issue for AST Fuzzer report https://github.com/ClickHouse/ClickHouse/issues/51486

**Review by serxa — APPROVED** [review]

**Comment by Felixoid** [comment]
Oh my.. the test actually caught the issue with the wrong leading slash. Cool

**Comment by Felixoid** [comment]
Ok, it's tested in cgroupsv2, but when we'll backport the changes to cloud releases, it should be there as well. So I am closing this PR in favor of https://github.com/ClickHouse/ClickHouse/pull/51180

**Pull request closed without merging**